src/packages/bem/README.md

```diff
@@ -0,0 +1,232 @@
+.ComponentName__elem--modName {}
+/* component's root node class name with string/number modifyer + value applied */
+.ComponentName__elem--modName_modValue {}
+```
+
+Example of usage
+----------------
+
+### Statefull component
+
+Button.js
+```js
+import React, { Component } from 'react';
+import PropTypes from 'prop-types';
+// 1. Don't forget to import `bem` decorator
+import bem from 'bem';
+// 2. Import classnames map (webpack's css loader should be configured in css-module mode)
+import styles from './Button.scss';
+
+// 3. Annotate component with `bem` decorator and pass styles to it
+@bem(styles)
+export default class Button extends Component {
+
+    static propTypes = {
+        active: PropTypes.bool,
+        type: PropTypes.oneOf([ 'normal', 'extraordinary' ]),
+    };
+    static defaultProps = {
+        active: false,
+        type: 'normal',
+    };
+
+    // 4. If you need to have class name (.Button--active) that depends on
+    //    `active` prop, just list this prop in propsToMods list.
+    static propsToMods = ['active'];
+
+    // 5. If you need to have class name (.Button--clicked) that depend on
+    //    state.clicked value, just list this prop in propsToMods list.
+    static stateToMods = ['clicked'];
+
+    state = { clicked: false };
+
+    handleClick = () => {
+        const { clicked } = this.state;
+        this.setState({ clicked: !clicked });
+    };
+
+    render() {
+        return (
+            {/*  6. Add { ...this.block() } construction to declare node as a block root */}
+            <button { ...this.block() } onClick={this.handleClick}>
+                {/*  7. Add { ...this.elem('label') } construction to declare node as a label element */}
+                <span { ...this.elem('label') }>
+                    {this.props.children}
+                </span>
+            </button>
+        );
+    }
+}
+```
+
+ButtonStateless.js
+```js
+import React, { Component } from 'react';
+import PropTypes from 'prop-types';
+import bem from 'bem';
+import styles from './Button.scss';
+
+// 3. Annotate component with `bem` decorator and pass styles to it
+import React from 'react';
+
+function ButtonStateless(props) {
+    return (
+      {/*  1. Add { ...props.block() } construction to declare node as a block */}
+      <button { ...props.block() }>
+        {/*  2. Add { ...props.elem('label') } construction to declare node as a label element */}
+        <span { ...props.elem('label') }>
+            {props.children}
+        </span>
+      </button>
+    );
+);
+
+ButtonStateless.displayName = 'Button';
+ButtonStateless.propTypes = {
+    active: PropTypes.bool,
+};
+ButtonStateless.defaultProps = {
+    active: false,
+};
+// 3. If you need to have class name (.ButtonStateless--active) that depends on
+//    `active` prop, just list this prop in propsToMods list.
+ButtonStateless.propsToMods = ['active'];
+
+export default bem(styles)(ButtonStateless);
```

Comment:
> Could we just use the HOC version without the decorator (as cool as it is) for class components too?

Technically, we can

> It makes refactoring as well as reading the code a lot more complicated.

I don't agree on that one, or might be I don't see what kind of refactoring are you talking about.
From my point of view component is more readable when you see all the important staff (like props declaration,  props defaults, decoration...) on the top of the file before implementation, not in the bottom.

In case of stateless function all that stuff goes after function body only because there is no other way to attach static stuff to a function in JS, not because it's a good or readable.

> I'm not so happy with using different conventions for stateful and stateless components.

Stateful and stateless components are different types of components, they uses different syntax, as for me it's ok.

Reply:
the main refactor when it is problematic is making a stateful to stateless component or the other way around.
The way I read the code, it normally makes little different when trying to understand thing if they are stateful or stateless. Hence the more the syntax is similar, the easier it is to read.

Reply:
Putting here my 5 cents, I don't have a strong opinion on this. For me, both ways are clean and easy to understand as long as we use our code format preferences.

But, if it's the case, I wouldn't mind making the syntax more similar, putting the `static` values of that stateful component outside itself, just to make it consistent with the stateless component. Although I don't see any problem with them inside the stateful component too :)

Reply:
In that case, I'll be happy to go with the more similar approach. In my experience making stateful into stateless and vv is a common thing, and it is painful enough with renaming all the `this` and method declaration without the need to move the props as well.

Reply:
Ok than, as I said nothing stops us from doing so